metricbeat/module/vsphere/resourcepool/resourcepool.go

```diff
@@ -0,0 +1,230 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package resourcepool
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/performance"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/view"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/vsphere"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each resourcepool is defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("vsphere", "resourcepool", New,
+		mb.WithHostParser(vsphere.HostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet type defines all fields of the MetricSet.
+type MetricSet struct {
+	*vsphere.MetricSet
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	ms, err := vsphere.NewMetricSet(base)
+	if err != nil {
+		return nil, err
+	}
+	return &MetricSet{ms}, nil
+}
+
+// Structure to hold performance metrics values
+type metricData struct {
+	perfMetrics map[string]interface{}
+	assetsName  assetNames
+}
+
+type assetNames struct {
+	outputVmNames []string
+}
+
+// Define metrics to be collected
+var metricNames = []string{
+	"cpu.usage.average",
+	"rescpu.actav1.latest",
+	"rescpu.actpk1.latest",
+	"cpu.usagemhz.average",
+	"mem.usage.average",
+	"mem.shared.average",
+	"mem.swapin.average",
+	"cpu.cpuentitlement.latest",
+	"mem.mementitlement.latest",
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
```

Comment:
Fetch method is quite large. It could be broken down in a few different functionalities.

Reply:
+1

Reply:
I guess current method is okay now as we have removed perfMetrics.